### PR TITLE
fix: Make using Java's ServiceLoader optional on Database.connect()

### PR DESCRIPTION
**Pull request by makeevrserg**
#### Description

**Summary of the change**: Make using ServiceLoader for DatabaseConnectionAutoRegistrationnot optional

**Detailed description**:
- **What**: `connectionInstanceImpl` doesn't need to be created using java `ServiceLoader`. 
- **Why**: In some cases with different loaders it may throw exception even if existing interface implementation is present. Also this may improve stability as reflection will not be required during connection.
- **How**: The `connectionInstanceImpl` is now lazy-loaded so even if `ServiceLoader` doesn't have required service, code block will not crash. Every `connect` method have now `connectionAutoRegistration` as a parameter instead of previous private-service-loaded
- This also relevant to older issues on github, which is marked as closed, but still could be reproduced located below
- This also may be a small step in migration to KMP if it's planned.

---

#### Type of Change

Please mark the relevant options with an "X":
- [x] Bug fix
- [x] New feature
- [x] Documentation update

Updates/remove existing public API methods:
- [x] Is breaking change

Affected databases:
- [ ] MariaDB
- [ ] Mysql5
- [ ] Mysql8
- [ ] Oracle
- [ ] Postgres
- [ ] SqlServer
- [ ] H2
- [ ] SQLite

#### Checklist

- [x] Unit tests are in place
- [x] The build is green (including the Detekt check)
- [x] All public methods affected by my PR has up to date API docs
- [x] Documentation for my change is up to date

---

#### Related Issues

- https://github.com/JetBrains/Exposed/issues/1353
- https://github.com/JetBrains/Exposed/issues/971
- https://github.com/JetBrains/Exposed/issues/792
- https://github.com/JetBrains/Exposed/issues/748

### CODE REVIEW

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Database.kt [hunk 1]
```diff
@@ -180,6 +183,7 @@ class Database private constructor(
          * but instead provides the details necessary to do so whenever a connection is required by a transaction.
          *
          * @param datasource The [DataSource] object to be used as a means of getting a connection.
+         * @param connectionAutoRegistration The connection provider for database. If not provided, will use ServiceLoaded [connectionInstanceImpl]
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Database.kt:186]: `connectionInstanceImpl` is a private property that may not mean anything to users.
Please change this (and the other KDocs below) to something more general, like:
```
* @param connectionAutoRegistration The connection provider for database. If not provided,
* a service loader will be used to locate and load a provider for [DatabaseConnectionAutoRegistration].
```

**Review by bog-walk — CHANGES_REQUESTED** [review]
Thanks for working on this PR @makeevrserg !
Please consider rebasing from `main` and addressing the suggested changes.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Database.kt (hunk 1 above)
Comment (makeevrserg) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Database.kt]: Fixed 🫡

File: gradle.properties [hunk 2]
```diff
@@ -4,4 +4,4 @@ org.gradle.configuration.cache=true
 org.gradle.caching=true
 
 group=org.jetbrains.exposed
-version=0.56.0
+version=0.57.0
```
[the remark this answers is not in the document]
Comment (makeevrserg) [reply gradle.properties]: Fixed 🫡

File: documentation-website/Writerside/topics/Working-with-Database.md [hunk 3]
```diff
@@ -12,6 +12,9 @@ Every database access using Exposed is started by obtaining a connection and cre
 First of all, you have to tell Exposed how to connect to a database by using the `Database.connect` function.
 It won't create a real database connection but will only provide a descriptor for future usage.
 
+By default, Exposed using `ServiceLoader` to get `DatabaseConnectionAutoRegistration`.
+It can be modified when calling `Database.connect` method by providing `connectionAutoRegistration` in parameter list.
+
```
[the remark this answers is not in the document]
Comment (makeevrserg) [reply documentation-website/Writerside/topics/Working-with-Database.md]: Fixed 🫡

File: documentation-website/Writerside/topics/Getting-Started-with-Exposed.topic [hunk 4]
```diff
@@ -194,6 +194,10 @@
             Every database access using Exposed is started by obtaining a connection and creating a transaction.
             To configure the database connection, use the <code>Database.connect()</code> function.
         </p>
+        <p>
+            By default, Exposed using `ServiceLoader` to get `DatabaseConnectionAutoRegistration`.
+            It can be modified when calling `Database.connect` method by providing `connectionAutoRegistration` in parameter list.
+        </p>
```
[the remark this answers is not in the document]
Comment (makeevrserg) [reply documentation-website/Writerside/topics/Getting-Started-with-Exposed.topic]: Fixed 🫡

**Review by e5l — APPROVED** [review]
thanks for the contribution, lgtm!

**Review by bog-walk — APPROVED** [review]

**Pull request merged**